src/transformStateWithClones.js

```diff
@@ -7,7 +7,24 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    }
+
+    stateHistory.push({ ...currentState });
```

Comment:
Consider using a `switch` statement instead of multiple `if...else if` statements when dealing with a variable (`action.type`) that can have limited, well-defined values. This will make the code more readable and maintainable.